File: src/tux_driver.c
/*
 * Tux Droid - Driver
 * Copyright (C) 2008 C2ME Sa <Acness : remi.jocaille@c2me.be>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */

#include <stdio.h>
#include <string.h>

#include "log.h"
#include "tux_battery.h"
#include "tux_cmd_parser.h"
#include "tux_descriptor.h"
#include "tux_error.h"
#include "tux_eyes.h"
#include "tux_firmware.h"
#include "tux_hw_status.h"
#include "tux_id.h"
#include "tux_leds.h"
#include "tux_light.h"
#include "tux_mouth.h"
#include "tux_pong.h"
#include "tux_sound_flash.h"
#include "tux_sw_status.h"
#include "tux_user_inputs.h"
#include "tux_spinning.h"
#include "tux_usb.h"
#include "tux_flippers.h"
#include "tux_types.h"
#include "version.h"

static bool driver_started = false;
static simple_callback_t end_cycle_funct;
static simple_callback_t dongle_connected_funct;
static simple_callback_t dongle_disconnected_funct;

static void on_frame(const unsigned char *data);
static void on_rf_state(unsigned char state);
static void on_usb_connect(void);
static void on_usb_disconnect(void);
static void on_read_loop_cycle_complete(void);

void TuxDrv_ResetPositions(void);

/**
 * 31/08/2012 - Joël Maatteotti <sfuser: joelmatteitti>
 */
LIBEXPORT bool TuxDrv_Eyes_cmd_off()
{
	return tux_eyes_cmd_off();
}

/**
 * 31/08/2012 - Joël Maatteotti <sfuser: joelmatteitti>
 */
LIBEXPORT bool TuxDrv_Mouth_cmd_off()
{
	return tux_mouth_cmd_off();
}

/**
 * 31/08/2012 - Joël Maatteotti <sfuser: joelmatteitti>
 */
LIBEXPORT bool TuxDrv_Spinning_cmd_off()
{
	return tux_spinning_cmd_off();
}

/**
 * 31/08/2012 - Joël Maatteotti <sfuser: joelmatteitti>
 */
LIBEXPORT bool TuxDrv_Flippers_cmd_off()
{
	return tux_flippers_cmd_off();
}

/**
 * 31/08/2012 - Joël Maatteotti <sfuser: joelmatteitti>
 */
LIBEXPORT char *TuxDrv_SoundFlash_dump_descriptor(char *p)
{
	return tux_sound_flash_dump_descriptor(p);
}

/**
 * 31/08/2012 - Joël Maatteotti <sfuser: joelmatteitti>
 */
LIBEXPORT void TuxDrv_LightLevel_update()
{
	tux_light_update_level();
}


/**
 *
 */
LIBEXPORT void
TuxDrv_SetStatusCallback(event_callback_t funct)
{
    tux_sw_status_set_event_callback(funct);
}

/**
 *
 */
LIBEXPORT void
TuxDrv_SetEndCycleCallback(simple_callback_t funct)
{
    end_cycle_funct = funct;
}

/**
 *
 */
LIBEXPORT void
TuxDrv_SetDongleConnectedCallback(simple_callback_t funct)
{
    dongle_connected_funct = funct;
}

/**
 *
 */
LIBEXPORT void
TuxDrv_SetDongleDisconnectedCallback(simple_callback_t funct)
{
    dongle_disconnected_funct = funct;
}

/**
 *  Callback function on frame receiving.
 *  @param data 4 bytes array of status.
 */
static void
on_frame(const unsigned char *data)
{
    unsigned char header;
    int ret;

    header = data[0];
    ret = tux_hw_status_parse_frame(data);

    if (ret == -1)
    {
        log_warning("STATUS FRAME : %.2x %.2x %.2x %.2x",
            data[0],
            data[1],
            data[2],
            data[3]);
    }

    switch (header)
    {
    /* Descriptor frames */
    case FRAME_HEADER_VERSION:
        tux_firmware_update_version();
        break;
    case FRAME_HEADER_REVISION:
        tux_firmware_update_revision();
        break;
    case FRAME_HEADER_AUTHOR:
        tux_firmware_update_author();
        break;
    case FRAME_HEADER_SOUND_VAR:
        tux_sound_flash_update();
        break;
    /* Status frames */
    case FRAME_HEADER_PORTS:
        if (header == ret)
        {
            tux_spinning_update_direction();
            tux_mouth_update_position();
            tux_eyes_update_position();
        }
        break;
    case FRAME_HEADER_POSITION1:
        if (header == ret)
        {
            tux_flippers_update_movements_remaining();
            tux_mouth_update_movements_remaining();
            tux_eyes_update_movements_remaining();
        }
        break;
    case FRAME_HEADER_POSITION2:
        if (header == ret)
        {
            tux_flippers_update_position();
            tux_spinning_update_movements_remaining();
            tux_eyes_update_motor();
            tux_mouth_update_motor();
            tux_flippers_update_motor();
            tux_spinning_update_left_motor();
            tux_spinning_update_right_motor();
        }
        break;
    case FRAME_HEADER_SENSORS1:
        if (header == ret)
        {
            tux_sound_flash_update_general_play();
            tux_user_inputs_update_head_button();
            tux_user_inputs_update_left_wing_button();
            tux_user_inputs_update_right_wing_button();
        }
        break;
    case FRAME_HEADER_ID:
        tux_id_update_number();
        break;
    case FRAME_HEADER_IR:
        tux_user_inputs_init_time_RC5();
        break;
    case FRAME_HEADER_BATTERY:
        tux_user_inputs_update_charger_state();
        if (header == ret)
        {
            tux_battery_update_level();
        }
        break;
    case FRAME_HEADER_LIGHT:
        if (header == ret)
        {
            tux_light_update_level();
        }
        break;
    case FRAME_HEADER_LED:
        if (header == ret)
        {
            tux_leds_update_left();
            tux_leds_update_right();
        }
        break;
    /* Don't work ...
    case FRAME_HEADER_FLASH_PROG:
        if (header == ret)
        {
            tux_sound_flash_update_prog_current_track();
            tux_sound_flash_update_prog_last_track_size();
        }
        break;
    */
    case FRAME_HEADER_AUDIO:
        tux_sound_flash_update_flash_play();
        break;
    /*case FRAME_HEADER_PONG:
        tux_pong_update();
        break;*/
    }
}

/**
 *  Callback function on radio state changed.
 *  @param state state of the radio connection.
 */
static void
on_rf_state(unsigned char state)
{
    tux_sw_status_set_intvalue(SW_ID_RF_STATE, state, true);
    if (state)
    {
        tux_descriptor_get();
        /* For reinitilizing the remote control task */
        tux_user_inputs_init();
        /* Reset the body state */
        TuxDrv_ResetPositions();
    }
}

/**
 *  Callback function on fux dongle plugging.
 */
static void
on_usb_connect(void)
{
    data_frame wakeup_frame = {0xB6, 0xFF, 0x01, 0x00};

    tux_descriptor_init();
    tux_hw_status_init();
    tux_sw_status_init();
    tux_user_inputs_init();
    tux_cmd_parser_init();
    tux_sw_status_set_intvalue(SW_ID_DONGLE_PLUG, true, true);
    /* Waking up Tux Droid */
    tux_usb_send_to_tux(wakeup_frame);
    if (dongle_connected_funct)
    {
        dongle_connected_funct();
    }
    /* If default windows sound card is Tuxdroid-TTS then set
     * the default card to TuxDroid-Audio */
#ifdef WIN32
    tux_sound_flash_avoid_tts_default_sound_card();
#endif
}

/**
 *  Callback function on fux dongle unplugging.
 */
static void
on_usb_disconnect(void)
{
    tux_sw_status_set_intvalue(SW_ID_RF_STATE, false, true);
    tux_sw_status_set_intvalue(SW_ID_DONGLE_PLUG, false, true);
    if (dongle_disconnected_funct)
    {
        dongle_disconnected_funct();
    }
}

/**
 *  Callback function on end of a cycle of usb read.
 */
static void
on_read_loop_cycle_complete(void)
{
    tux_user_inputs_update_RC5();
    /* tux_pong_get(); */
    /* tux_firmware_state_machine_call(); */
    tux_sound_flash_state_machine_call();
    tux_hw_status_header_counter_check();

    if (end_cycle_funct)
    {
        end_cycle_funct();
    }

    tux_cmd_parser_delay_stack_perform();
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_PerformCommand(double delay, const char *cmd_str)
{
    if (delay == 0.0)
    {
        log_debug("Perform an instant command : [%s]", cmd_str);
        return tux_cmd_parser_parse_command(cmd_str);
    }
    else
    {
        log_debug("Perform a delayed command : [%s], %f", cmd_str, delay);
        return tux_cmd_parser_insert_user_command(delay, cmd_str);
    }
}

/**
 *
 */
LIBEXPORT void
TuxDrv_ClearCommandStack(void)
{
    tux_cmd_parser_clear_delay_commands();
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_PerformMacroFile(const char *file_path)
{
    return tux_cmd_parser_parse_file(file_path);
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_PerformMacroText(const char *macro)
{
    return tux_cmd_parser_parse_macro(macro);
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_SoundReflash(const char *tracks)
{
    return tux_sound_flash_cmd_reflash(tracks);
}

/**
 *
 */
LIBEXPORT void
TuxDrv_SetLogLevel(log_level_t level)
{
    log_set_level(level);
}

/**
 *
 */
LIBEXPORT void
TuxDrv_SetLogTarget(log_target_t target)
{
    log_open(target);
}

/**
 *
 */
LIBEXPORT int
TuxDrv_TokenizeStatus(char *status, tokens_t *tokens)
{
    return tux_cmd_parser_get_tokens(status, tokens, 32, ":,");
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_GetStatusName(int id, char *name)
{
    return tux_sw_status_name_from_id(id, name);
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_GetStatusId(const char *name, int *id)
{
    return tux_sw_status_id_from_name(name, id);
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_GetStatusState(int id, char *state)
{
    return tux_sw_status_get_state_str(id, state);
}

/**
 *
 */
LIBEXPORT void
TuxDrv_GetAllStatusState(char *state)
{
    tux_sw_status_get_all_state_str(state);
}

/**
 *
 */
LIBEXPORT TuxDrvError
TuxDrv_GetStatusValue(int id, char *value)
{
    return tux_sw_status_get_value_str(id, value);
}

/**
 *
 */
LIBEXPORT void
TuxDrv_ResetDongle(void)
{
    tux_usb_reset();
}

/**
 *
 */
LIBEXPORT void
TuxDrv_ResetPositions(void)
{
    TuxDrv_PerformCommand(0.0, "TUX_CMD:EYES:OPEN");
    TuxDrv_PerformCommand(0.3, "TUX_CMD:MOUTH:CLOSE");
    TuxDrv_PerformCommand(0.0, "TUX_CMD:FLIPPERS:DOWN");
    TuxDrv_PerformCommand(0.0, "TUX_CMD:SPINNING:OFF");
    TuxDrv_PerformCommand(0.0, "TUX_CMD:LED:ON:LED_BOTH,1.0");
}

/**
 *
 */
LIBEXPORT void
TuxDrv_GetDescriptor(tux_descriptor_t *tux_desc)
{
    tux_desc->firmwares.package = &firmware_release_desc;
    tux_desc->firmwares.tuxcore = &firmwares_desc[TUXCORE_CPU_NUM];
    tux_desc->firmwares.tuxaudio = &firmwares_desc[TUXAUDIO_CPU_NUM];
    tux_desc->firmwares.tuxrf = &firmwares_desc[TUXRF_CPU_NUM];
    tux_desc->firmwares.fuxrf = &firmwares_desc[FUXRF_CPU_NUM];
    tux_desc->firmwares.fuxusb = &firmwares_desc[FUXUSB_CPU_NUM];
    tux_desc->sound_flash = &sound_flash_desc;
    tux_desc->id = &id_desc;
    tux_desc->driver.version_major = VER_MAJOR;
    tux_desc->driver.version_minor = VER_MINOR;
    tux_desc->driver.version_update = VER_UPDATE;
    tux_desc->driver.version_build = VER_REVISION;
    strcpy(tux_desc->driver.version_state, VER_STATE);
    sprintf(tux_desc->driver.version_string,
        "libtuxdriver_%d.%d.%d-r%d",
        VER_MAJOR,
        VER_MINOR,
        VER_UPDATE,
        VER_REVISION);
}

/**
 * Get the description of an error code.
 */
LIBEXPORT const char *
TuxDrv_StrError(TuxDrvError error_code)
{
    return tux_error_strerror(error_code);
}

/**
 *  Start tux driver.
 */
LIBEXPORT void
TuxDrv_Start(void)
{
    printf("libtuxdriver_%d.%d.%d-r%d\n\n",
        VER_MAJOR,
        VER_MINOR,
        VER_UPDATE,
        VER_REVISION);

    tux_usb_init_module();
    tux_usb_set_frame_callback(on_frame);
    tux_usb_set_rf_state_callback(on_rf_state);
    tux_usb_set_connect_dongle_callback(on_usb_connect);
    tux_usb_set_disconnect_dongle_callback(on_usb_disconnect);
    tux_usb_set_loop_cycle_complete_callback(on_read_loop_cycle_complete);
    tux_descriptor_init();
    tux_hw_status_init();
    tux_sw_status_init();
    tux_user_inputs_init();
    tux_cmd_parser_init();
    driver_started = true;

    while (driver_started)
    {
        tux_usb_start();
        sleep(1);
    }

    tux_usb_exit_module();
}

/**
 * Stop tux driver.
 */
LIBEXPORT void
TuxDrv_Stop(void)
{
    driver_started = false;
    tux_usb_stop();
}

File: include/tux_driver.pas
{*
 * Tux Droid - Driver
 * Copyright (C) 2008 C2ME Sa <Acness : remi.jocaille@c2me.be>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *}
 
unit tux_driver;

interface

uses
  Windows, classes;
  
const

  DLL_NAME = 'libtuxdriver.dll';

  {**
   * Error codes enumeration.
   *}
  E_TUXDRV_BEGIN                    = 256;
  E_TUXDRV_NOERROR                  = 0;
  E_TUXDRV_PARSERISDISABLED         = E_TUXDRV_BEGIN;
  E_TUXDRV_INVALIDCOMMAND           = E_TUXDRV_BEGIN + 1;
  E_TUXDRV_STACKOVERFLOW            = E_TUXDRV_BEGIN + 2;
  E_TUXDRV_FILEERROR                = E_TUXDRV_BEGIN + 3;
  E_TUXDRV_BADWAVFILE               = E_TUXDRV_BEGIN + 4;
  E_TUXDRV_INVALIDIDENTIFIER        = E_TUXDRV_BEGIN + 5;
  E_TUXDRV_INVALIDNAME              = E_TUXDRV_BEGIN + 6;
  E_TUXDRV_INVALIDPARAMETER         = E_TUXDRV_BEGIN + 7;
  E_TUXDRV_BUSY                     = E_TUXDRV_BEGIN + 8;
  E_TUXDRV_WAVSIZEEXCEDED           = E_TUXDRV_BEGIN + 9;
    
  {** 
   * Id enumeration of high level status.
   *}
  SW_ID_FLIPPERS_POSITION           = 0;
  SW_ID_FLIPPERS_REMAINING_MVM      = 1;
  SW_ID_SPINNING_DIRECTION          = 2;
  SW_ID_SPINNING_REMAINING_MVM      = 3;
  SW_ID_LEFT_WING_BUTTON            = 4;
  SW_ID_RIGHT_WING_BUTTON           = 5;
  SW_ID_HEAD_BUTTON                 = 6;
  SW_ID_REMOTE_BUTTON               = 7;
  SW_ID_MOUTH_POSITION              = 8;
  SW_ID_MOUTH_REMAINING_MVM         = 9;
  SW_ID_EYES_POSITION               = 10;
  SW_ID_EYES_REMAINING_MVM          = 11;
  SW_ID_DESCRIPTOR_COMPLETE         = 12;
  SW_ID_RF_STATE                    = 13;
  SW_ID_DONGLE_PLUG                 = 14;
  SW_ID_CHARGER_STATE               = 15;
  SW_ID_BATTERY_LEVEL               = 16;
  SW_ID_BATTERY_STATE               = 17;
  SW_ID_LIGHT_LEVEL                 = 18;
  SW_ID_LEFT_LED_STATE              = 19;
  SW_ID_RIGHT_LED_STATE             = 20;
  SW_ID_CONNECTION_QUALITY          = 21;
  SW_ID_AUDIO_FLASH_PLAY            = 22;
  SW_ID_AUDIO_GENERAL_PLAY          = 23;
  SW_ID_FLASH_PROG_CURR_TRACK       = 24;
  SW_ID_FLASH_PROG_LAST_TRACK_SIZE  = 25;
  SW_ID_TUXCORE_SYMBOLIC_VERSION    = 26;
  SW_ID_TUXAUDIO_SYMBOLIC_VERSION   = 27;
  SW_ID_FUXUSB_SYMBOLIC_VERSION     = 28;
  SW_ID_FUXRF_SYMBOLIC_VERSION      = 29;
  SW_ID_TUXRF_SYMBOLIC_VERSION      = 30;
  SW_ID_DRIVER_SYMBOLIC_VERSION     = 31;
  SW_ID_SOUND_REFLASH_BEGIN         = 32;
  SW_ID_SOUND_REFLASH_END           = 33;
  SW_ID_SOUND_REFLASH_CURRENT_TRACK = 34;
  SW_ID_EYES_MOTOR_ON               = 35;
  SW_ID_MOUTH_MOTOR_ON              = 36;
  SW_ID_FLIPPERS_MOTOR_ON           = 37;
  SW_ID_SPIN_LEFT_MOTOR_ON          = 38;
  SW_ID_SPIN_RIGHT_MOTOR_ON         = 39;
  SW_ID_FLASH_SOUND_COUNT           = 40;

type

  {** 
   * Simple callback definition.
   *}
  drv_simple_callback_t = procedure;

  {**
   * Status callback definition.
   *}
  drv_status_callback_t = procedure(status:PChar); cdecl;
  
  {** 
   * Logging target
   *}
  log_target_t = (
    LOG_TARGET_TUX = 0,
    LOG_TARGET_SHELL
  );

  {** 
   * Logging levels, in increasing priorities 
   *}
  log_level_t = (
    LOG_LEVEL_DEBUG = 0,
    LOG_LEVEL_INFO,
    LOG_LEVEL_WARNING,
    LOG_LEVEL_ERROR,
    LOG_LEVEL_NONE
  );
  
  {** 
   * Descriptor structure of a firmaware.
   *}
  p_firmware_descriptor_t = ^firmware_descriptor_t;
  firmware_descriptor_t = packed record
    cpu_id              : integer;
    version_major       : cardinal;
    version_minor       : cardinal;
    version_update      : cardinal;
    revision            : cardinal;
    rlmmr               : cardinal;
    author              : cardinal;
    variation           : cardinal;
    version_string      : packed array[0..255] of char;
  end;

  {** 
   * Descriptor structure of sound flash.
   *}
  p_sound_flash_descriptor_t = ^sound_flash_descriptor_t;
  sound_flash_descriptor_t = packed record
    number_of_sounds      : cardinal;
    flash_usage           : cardinal;
    available_record_time : cardinal;
  end;
  
  {** 
   * Descriptor structure of ID connection.
   *}
  p_id_descriptor_t = ^id_descriptor_t;
  id_descriptor_t = packed record
    number    : cardinal;
  end;
  
  {** 
   * Global descriptor structure of tuxdroid.
   *}
  p_tux_descriptor_t = ^tux_descriptor_t;
  tux_descriptor_t = packed record
    firmwares : packed record
      package         : p_firmware_descriptor_t;
      tuxcore         : p_firmware_descriptor_t;
      tuxaudio        : p_firmware_descriptor_t;
      tuxrf           : p_firmware_descriptor_t;
      fuxrf           : p_firmware_descriptor_t;
      fuxusb          : p_firmware_descriptor_t;
    end;
    driver : packed record
      version_major   : cardinal;
      version_minor   : cardinal;
      version_update  : cardinal;
      version_build   : cardinal;
      version_state   : packed array[0..99] of char;
      version_string  : packed array[0..99] of char;
    end;
    sound_flash       : p_sound_flash_descriptor_t;
    id                : p_id_descriptor_t;
  end;

{**
 * DLL static linkage.
 *}
procedure TuxDrv_Start; cdecl;
  external DLL_NAME name 'TuxDrv_Start';  
procedure TuxDrv_Stop; cdecl;
  external DLL_NAME name 'TuxDrv_Stop';  
procedure TuxDrv_SetStatusCallback(funct:drv_status_callback_t); cdecl;
  external DLL_NAME name 'TuxDrv_SetStatusCallback';   
procedure TuxDrv_SetEndCycleCallback(funct:drv_simple_callback_t); cdecl;
  external DLL_NAME name 'TuxDrv_SetEndCycleCallback';   
procedure TuxDrv_SetDongleConnectedCallback(funct:drv_simple_callback_t); cdecl;
  external DLL_NAME name 'TuxDrv_SetDongleConnectedCallback';   
procedure TuxDrv_SetDongleDisconnectedCallback(funct:drv_simple_callback_t); cdecl;
  external DLL_NAME name 'TuxDrv_SetDongleDisconnectedCallback';   
function TuxDrv_PerformCommand(delay:real; cmd_str:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_PerformCommand';  
function TuxDrv_PerformMacroFile(file_path:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_PerformMacroFile';  
function TuxDrv_PerformMacroText(macro:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_PerformMacroText';  
function TuxDrv_GetStatusName(id:integer; name:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_GetStatusName';  
function TuxDrv_GetStatusValue(id:integer; value:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_GetStatusValue';  
function TuxDrv_GetStatusId(name:PChar; var id:integer):integer; cdecl;
  external DLL_NAME name 'TuxDrv_GetStatusId';  
procedure TuxDrv_ClearCommandStack; cdecl;
  external DLL_NAME name 'TuxDrv_ClearCommandStack';  
function TuxDrv_GetStatusState(id:integer; state:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_GetStatusState';  
procedure TuxDrv_GetAllStatusState(state:PChar); cdecl;
  external DLL_NAME name 'TuxDrv_GetAllStatusState';  
function TuxDrv_SoundReflash(tracks:PChar):integer; cdecl;
  external DLL_NAME name 'TuxDrv_SoundReflash';    
procedure TuxDrv_ResetPositions; cdecl;
  external DLL_NAME name 'TuxDrv_ResetPositions';  
procedure TuxDrv_ResetDongle; cdecl;
  external DLL_NAME name 'TuxDrv_ResetDongle';  
procedure TuxDrv_SetLogLevel(level:integer); cdecl;
  external DLL_NAME name 'TuxDrv_SetLogLevel';  
procedure TuxDrv_SetLogTarget(target:integer); cdecl;
  external DLL_NAME name 'TuxDrv_SetLogTarget'; 
procedure TuxDrv_GetDescriptor(tux_desc:p_tux_descriptor_t); cdecl;  
  external DLL_NAME name 'TuxDrv_GetDescriptor';  
function TuxDrv_StrError(error_code:integer):PChar; cdecl;
  external DLL_NAME name 'TuxDrv_StrError';
  
implementation

end.
